client/src/app/gateways/vote-decrypt-gateway.service.ts

```diff
@@ -0,0 +1,82 @@
+import { Injectable } from '@angular/core';
+import { getPublicKey, getSharedSecret, utils, verify } from '@noble/ed25519'
+
+import { ViewPoll } from '../site/pages/meetings/pages/polls';
+import { ORGANIZATION_ID } from '../site/pages/organization/services/organization.service';
+import { OrganizationControllerService } from '../site/pages/organization/services/organization-controller.service';
+
+function uInt8Enc(str: string, decode = false): Uint8Array {
+    const useStr = decode? atob(str) : str;
+    return Uint8Array.from(useStr, c => c.charCodeAt(0));
+}
+
+@Injectable({
+  providedIn: `root`
+})
+export class VoteDecryptGatewayService {
+    private _publicMainKey: Uint8Array;
+    private readonly _publicKeySize: number = 32;
+    private readonly _nonceSize: number = 12;
+
+    constructor(
+        private orgaController: OrganizationControllerService
+    ) {
+        this.orgaController.getViewModelObservable(ORGANIZATION_ID).subscribe(organization => {
+            this._publicMainKey = uInt8Enc(organization?.vote_decrypt_public_main_key, true);
+            console.log(`NEW publicMainKey: `, this._publicMainKey);
+        });
+    }
+
+    public async encryptVote(poll: ViewPoll, vote: string): Promise<string> {
+        console.log(`----------[encryptVote]----------`);
+        const plaintext = new TextEncoder().encode(vote);
+        const cryptKey = uInt8Enc(poll.crypt_key, true);
+        const cryptSignature = uInt8Enc(poll.crypt_signature, true);
+        console.log(`transformed: \n>plaintext: `, plaintext, `\n>cryptKey: `, cryptKey, `\n>cryptSignature: `, cryptSignature);
+
+        if (!(await verify(cryptSignature, cryptKey, this._publicMainKey))) {
+            throw new Error(`Voting failed: Cryptography keys could not be verified.`);
+        }
+
+        const { ephPrivKey, ephPublKey, sharedSecret } = await this.generateKeysAndSecret(cryptKey)
+        console.log(`>ephPrivKey: `, ephPrivKey, `\n>ephPublKey: `, ephPublKey, `\n>sharedSecret: `, sharedSecret);
+
+        const nonce = crypto.getRandomValues(new Uint8Array(this._nonceSize));
+        console.log(`nonce: `, nonce);
+
+        const hkdfKey = await this.generateHkdfKey(sharedSecret)
+        console.log(`hkdfKey: `, hkdfKey);
+
+        const encrypted = new Uint8Array(await (crypto.subtle.encrypt({name: `AES-GCM`, iv: nonce}, hkdfKey, plaintext) as Promise<ArrayBuffer>));
+        console.log(`encrypted: `, encrypted);
+
+        return this.createPayload(ephPublKey, nonce, encrypted);
+    }
+
+    private async generateKeysAndSecret(cryptKey: Uint8Array): Promise<{ephPrivKey: Uint8Array, ephPublKey: Uint8Array, sharedSecret: Uint8Array}> {
```

Comment:
The private key does not have to be returned. Only the public key and the sharedSecret are needed